pkg/operator/api/v1alpha1/gameserver_types.go

```diff
@@ -80,6 +80,8 @@ type GameServerStatus struct {
 	Health GameServerHealth `json:"health,omitempty"`
 	// State defines the state of the game server (Initializing, StandingBy, Active etc.)
 	State GameServerState `json:"state,omitempty"`
+	// The Previously known manually set state
```

Comment:
I'm surprised the installfiles are not modified by this addition. When you run `make` and/or `make manifests` inside the `pkg/operator` folder, do the CRD YAML files get modified?

Reply:
You're right, it does add a new field to the "mps.playfab.com_gameservers.yaml" in pkg/operator/config/crd/bases. I'll have to find a different approach rather that doesn't trigger the "installfiles" check rather than the new "prevState" field

Reply:
thanks - I mean, if we *have to* modify the install files, that's ok - it's not a breaking change since old clients will not break (it would be a real issue if we removed a field). Up to you, thanks!

Reply:
So, I've tried a few different ways. Such as exporting metrics from the nodeagent and gameserver_controller. For high-level controller metrics, the gameserverbuild_controller, which is already managing, iterating through and counting every other GameServer seems like the best method. It handles the actual deleting and creating of gameservers, so unlike the nodeagents and gameserver_controller, it's already _mostly_ aware of everything occurring relating to servers

However, the only thing keeping it from producing more desired metrics than what it is currently capable of, is the fact that it isn't aware of state transitions, only current states. So, I do think this one field addition would be helpful (and as you mentioned, it's a safe API addition).

Reply:
Apologies for the back and forth, but I was just thinking that we `kinda` know what the previous state was, right? We only need to calculate the time diff. What if we add these fields to the .Status:

- timeWhenInitializing
- timeWhenStandingBy
- timeWhenActive

All the aforementioned fields will contain the UTC time that the state transition took place. The first two states are updated via the NodeAgent where the third is updated by the allocation API service.

I'm not sure if we can emit metrics this way (maybe we can, from the NodeAgent?) but it can definitely help user diagnose issues.
Sorry, I understand that this will require re-coding some stuff but it looks like an interesting solution. Anxious to hear your thoughts, thanks again for working on that!

Reply:
Inspiration for this idea is the Pod.Status.Conditions array which contains timestamps for PodPhase changes.

Reply:
That's a good idea for fault-tolerant time tracking. 

The major issue I ran into with the NodeAgent and the gameserver_controller was handling scale down metrics as the NodeAgent and gameserver_controller both have ways of responding to a gameserver that was deleted, but usually only after it has happened, while the gameserverbuild_controller actually triggers the cleanups.

However, on the other hand, the issue I ran in with the gameserverbuild_controller was that unless it has at least one field in the gameserver that it controls (such as updating the previous state), then there isn't really a clean way to distinguish between pods that have already had their states factored into the reconcile performance metrics and pods that have not.

Without knowing this, then rather than having performance metrics associated with reconcile loops, the metrics would have to be a running average of all existing pods, or just a flat display of the reconcile performance metrics associated with the last managed gameserver.

So if the NodeAgent or gameserver_controller are used, the scale down metrics would still need to be in the gameserverbuild_controller.

P.S. I appreciate the back and forth, there are a lot of variables to this

Reply:
Love the discussion, so many great points and loose variables! Let's take a step back and discuss the scenarios we want to tackle.

### Scaling up

We scale up our GameServerBuild when  we initially create the GameServerBuild and when we want to increase the StandingBy count for increased user load. 
It is crucial to understand the timings of the state transitions to find potential regressions and to ensure our matchmaker can start sending users only when the required StandingBy levels have been reached.

### Scaling down

We scale down our GameServerBuild when we want to save $$$ and the remaining StandingBy levels can satisfy the demand and when we want to upgrade our game server version (https://playfab.github.io/thundernetes/howtos/upgradebuild.html). The timings here (when the Pod is deleted) are usually less important than the "scaling up" ones. Once we're done with a GameServer, (hopefully) all the client connections will have been dropped so apart from some potential cleanup code, we do not have to wait or worry about anything else. At the same time, by having an extra .Patch for each GameServer in the GameServerBuild controller, we'd have to deal with an extra GameServer reconciliation loop, making the system *slightly* less performant.

Consequently, I'd recommend that as a first step we focus on the scale up metrics which seems easier to tackle. What do you think? Thanks for a great discussion!

Reply:
So, I have to apologize since I just saw that the GameServer status already contains timestamps for when it goes to the various states (Initializing/StandingBy/Active)

https://github.com/PlayFab/thundernetes/blob/main/pkg/operator/api/v1alpha1/gameserver_types.go#L97

These are set by the NodeAgent so I'd assume it's pretty trivial to create metrics based on them. WDYT? Thanks again and apologies for the inconvenience!

Reply:
Oh no worries, that's what I ended up doing. I didn't have time to perform any testing Friday and didn't want to update the PR until after. Additionally, I had to add a proper timestamp to the GameServerInfo for finding the difference from the existing StandByTime field; the creation time currently in-place is only an integer for finding the duration with the heartbeat.

Reply:
Thank you, much appreciate and apologies again for the back and forth!
I left a comment for the CreationTime, I *think* we can use that. Let me know!